[quant][pyper] Support aten::embedding_bag quantization in graph mode

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44048 [quant][pyper] Support quantization of ops in fork-wait subgraph
* #44008 [quant][pyper] make embedding_bag quantization static
* **#43989 [quant][pyper] Support aten::embedding_bag quantization in graph mode**

Summary:
When we trace the model it produces aten::embedding_bag node in the graph,
Add necessary passes in graph mode to help support quantizing it as well

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23460485](https://our.internmc.facebook.com/intern/diff/D23460485)

supriyar committed:
[quant] Support aten::embedding_bag quantization in graph mode
Summary:
When we trace the model it produces aten::embedding_bag node in the graph,
Add necessary passes in graph mode to help support quantizing it as well

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] Support aten::embedding_bag quantization in graph mode"


Summary:
When we trace the model it produces aten::embedding_bag node in the graph,
Add necessary passes in graph mode to help support quantizing it as well

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23460485](https://our.internmc.facebook.com/intern/diff/D23460485)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][pyper] Support aten::embedding_bag quantization in graph mode"


Summary:
When we trace the model it produces aten::embedding_bag node in the graph,
Add necessary passes in graph mode to help support quantizing it as well

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23460485](https://our.internmc.facebook.com/intern/diff/D23460485)

[ghstack-poisoned]

vkuzo commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -259,10 +260,16 @@ bool matchArgPattern(
       v,
-      AtenFuncArgs(
-          {{"conv1d", 1}, {"conv2d", 1}, {"conv3d", 1}, {"linear", 1}}),
+      // ate::embedding_bag(%weight, %input, %offsets, %scale_grad_by_freq,
```
nit: `aten`?

vkuzo reviewed (approved):
lg, feel free to wait for @jerryzh168  if a deeper review is needed

raghuramank100 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2917,21 +2919,26 @@ def forward(self, indices1, offsets1, indices2, offsets2):
+            from torch.quantization import QConfigDynamic, PlaceholderObserver
+            int4_dynamic_qconfig = QConfigDynamic(activation=PlaceholderObserver.with_args(dtype=torch.float,
+                                                                                           custom_op_name="embedding_bag_4bit"),
+                                                  weight=PlaceholderObserver.with_args(custom_op_name="embedding_bag_4bit"))
```
Why do we have a placeholder observer for weights?

supriyar replied:
We currently use real observers and torchbind classes for eager mode 8-bit embedding quant currently. For graph mode we implemented this initially using the custom prepack ops for PyPer for 8bit and 4bit, to be consistent with C2.
Going forward, in fx we can implement embeddingbag quantization using observers. I feel it is a bit of an overkill to update this code to use observers for 8-bit and placeholder observers for 4-bit. Let me know your thoughts.

supriyar committed:
Update on "[quant][pyper] Support aten::embedding_bag quantization in graph mode"


Summary:
When we trace the model it produces aten::embedding_bag node in the graph,
Add necessary passes in graph mode to help support quantizing it as well

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23460485](https://our.internmc.facebook.com/intern/diff/D23460485)

[ghstack-poisoned]
